### PR TITLE
remember last error and keep internal buffer of last N errors. fixed …

### LINKED ISSUE #45: Store history of last N errors and api to get last error or all history
We want to use circuit breaker instance in our healthcheck reporting. Would be nice if cb instance was able to report which errors triggered it to fail.
If that's something you consider useful for this library we can send a PR.

### PR DESCRIPTION
**Pull request by Kane-Sendgrid**
…unreliable timeout test

Fixes #45

### CODE REVIEW

File: circuitbreaker.go [hunk 1]
```diff
@@ -153,12 +159,17 @@ func NewBreakerWithOptions(options *Options) *Breaker {
 		options.WindowBuckets = DefaultWindowBuckets
 	}
 
+	if options.ErrorHistoryDepth == 0 {
```
Comment (dtjm) [inline circuitbreaker.go]: what if it's < 0?  Should this be uint?

File: circuitbreaker.go [hunk 2]
```diff
@@ -293,6 +304,31 @@ func (cb *Breaker) Fail() {
...
 
+// FailWithError is the same as Fail, but keeps history of errors in internal ring buffer
+func (cb *Breaker) FailWithError(err error) {
+	cb.errorsBuffer = cb.errorsBuffer.Next()
+	cb.errorsBuffer.Value = err
+	cb.Fail()
+}
+
+// Error returns last error from internal buffer
+func (cb *Breaker) Error() error {
```
Comment (dtjm) [inline circuitbreaker.go]: should return `string` to conform to `error` interface

File: circuitbreaker.go [hunk 3]
```diff
@@ -293,6 +304,31 @@ func (cb *Breaker) Fail() {
...
+	if cb.errorsBuffer.Value == nil {
+		return nil
+	}
+	return cb.errorsBuffer.Value.(error)
+}
+
+// Errors returns all errors from internal buffer
+func (cb *Breaker) Errors() (errors []error) {
+	cb.errorsBuffer.Do(func(x interface{}) {
```
Comment (dtjm) [inline circuitbreaker.go]: This will output the current error and then the rest.

errors ordered as `1 2 3` will be output as `3 1 2`

File: circuitbreaker.go [hunk 4]
```diff
@@ -293,6 +304,31 @@ func (cb *Breaker) Fail() {
...
+		if x != nil {
+			errors = append(errors, x.(error))
```
Comment (dtjm) [inline circuitbreaker.go]: might as well pre-allocate slice since you know the max number of elements

File: circuitbreaker.go [hunk 5]
```diff
@@ -302,7 +338,7 @@ func (cb *Breaker) Success() {
 	cb.backoffLock.Unlock()
 
 	state := cb.state()
-	if state == halfopen {
+	if state == halfopen || state == open {
```
Comment (dtjm) [inline circuitbreaker.go]: add comment why this is needed?

File: circuitbreaker_test.go [hunk 6]
```diff
@@ -253,6 +253,48 @@ func TestThresholdBreakerCalling(t *testing.T) {
...
+	if len(errs) != 2 {
+		t.Fatalf("expected `%d` errors, got %d", 2, len(errs))
+	}
+	if errs[0].Error() != "circuit error 1" {
+		t.Fatalf("expected `%s` error, got %s", "circuit error 1", errs[0].Error())
+	}
+	if errs[1].Error() != "circuit error" {
+		t.Fatalf("expected `%s` error, got %s", "circuit error", errs[0].Error())
+	}
+}
```
Comment (dtjm) [inline circuitbreaker_test.go]: add test with empty error buffer

**Comment by dtjm** [comment]
Add test for race?

**Comment by dtjm** [comment]
@rubyist can this be merged?
